server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -738,8 +738,8 @@ public Object readGenericValue() throws IOException {
             case 6 -> readByteArray();
             case 7 -> readArrayList();
             case 8 -> readArray();
-            case 9 -> readLinkedHashMap();
-            case 10 -> readHashMap();
+            case 9 -> version.onOrAfter(Version.V_8_4_0) ? readObjectKeyLinkedHashMap() : readLinkedHashMap();
+            case 10 -> version.onOrAfter(Version.V_8_4_0) ? readObjectKeyHashMap() : readHashMap();
```

Comment:
I think maybe we should modify `readLinkedHashMap` to do the object keys. So the version test would be inside those methods.

Reply:
Ignore this comment - Ryan proposed a better solution using the generic map readers.